### PR TITLE
set discoveryAddress to be cluster-local by default

### PR DESCRIPTION
**Pull request by stevenctl**
Any Envoy trying to connect to a discovery server will either:
- have an istiod instance in the same cluster
- have an istiod-remote service that forwards traffic to the primary cluster's ingressgateway

In either case, traffic from the proxy to the discovery server should be cluster-local. 

This fixes https://github.com/istio/istio/issues/23591

### CODE REVIEW

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ import (
 	"strings"
 	"time"
 
+	"istio.io/istio/pkg/config/host"
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/context.go:28]: move down

**Comment by stevenctl** [comment]
@hzxuzhonghu 
>One question: should we make istio-system all cluster local as kube-system?


Created https://github.com/istio/istio/pull/23642 to see if doing so fails any existing integration tests. Don't see why it would, wondering if there is a case where we would want to route straight to the istiod service via NodePort instead of istiod-remote->ingressgateway. I don't think anyone is doing that in prod but I feel like I've heard it mentioned.

**Comment by stevenctl** [comment]
I'm going to rebase on https://github.com/istio/istio/pull/23208, the `multicluster/remote` suite is what helped me discover the bug. Want to make sure that test would pass with this change.

File: pilot/pkg/model/push_context_test.go [hunk 2]
```diff
@@ -461,6 +461,18 @@ func TestIsClusterLocal(t *testing.T) {
...
 			expected: true,
 		},
+		{
+			name:     "discovery server is local by default",
+			m:        mesh.DefaultMeshConfig(),
+			host:     "istiod.istio-system.svc",
+			expected: true,
+		},
+		{
+			name:     "discovery server is local with domain suffix",
```
Comment (stevenctl) [inline pilot/pkg/model/push_context_test.go:471]: @nmittler when checking "IsClusterLocal" would we expect that they will always include the domain suffix?

Comment (nmittler) [reply pilot/pkg/model/push_context_test.go:471]: I believe so, are you not seeing that?

Comment (stevenctl) [reply pilot/pkg/model/push_context_test.go:471]: I'm only seeing that, I was just seeing if we need to be robust to both with and without.

Comment (nmittler) [reply pilot/pkg/model/push_context_test.go:471]: No, I think everything should be fqdn, so we're good.

**Comment by costinm** [comment]
I'm not sure this is correct - it may be an option that some users may want, but there are plenty of cases where you want envoys to pick control plane from remote clusters even if a local control plane is running.  The point of having control plane in multiple clusters is to allow workloads to load balance and survive failure of any of the control planes.

**Comment by stevenctl** [comment]
@costinm 
> I'm not sure this is correct - it may be an option that some users may want, but there are plenty of cases where you want envoys to pick control plane from remote clusters even if a local control plane is running. The point of having control plane in multiple clusters is to allow workloads to load balance and survive failure of any of the control planes.

In that case maybe it would be better to advise users to set the cluster local up themselves in mesh config or create a profile to set this up.

**Comment by nmittler** [comment]
@costinm 

> I'm not sure this is correct - it may be an option that some users may want, but there are plenty of cases where you want envoys to pick control plane from remote clusters even if a local control plane is running. The point of having control plane in multiple clusters is to allow workloads to load balance and survive failure of any of the control planes.

This is the first I've heard that we support something like this. Are you aware of anyone doing this today? Do we have it documented anywhere?

Master/remote is broken without this, so if we back this out we need an alternative.

FYI @howardjohn @smawson @linsun

**Comment by smawson** [comment]
I think it is a not-well-advertised feature (maybe not documented at all?). So far people have been more people seem surprised by this behavior than expecting it. I think we should make the default be cluster-local for everything in istio-system and carefully add the ability for users to use control plane replication if they want it. We also need to figure out how to separate out different services if we're running a single service with full support in some clusters and partial support in others, those shouldn't be the same k8s service name or they should not be merged.

**Comment by linsun** [comment]
@stevenctl 
```
Any Envoy trying to connect to a discovery server will either:
have an istiod instance in the same cluster
have an istiod-remote service that forwards traffic to the primary cluster's ingressgateway
```
this is not entirely true
it can also be: not have an istiod instance in the same cluster but have an istiod service that foward traffic to primary cluster’s ingress gw

**Pull request merged**